src/ctia/entity/event.clj

```diff
@@ -62,22 +69,89 @@
 
 (def EventGetParams EventFieldsParam)
 
+(defn same-bucket? [bucket event]
+  (let [max-seconds (get-in @properties [:ctia :events :timeline :max-seconds] 5)
+        from        (t/minus (:from bucket) (t/seconds max-seconds))
+        to          (t/plus (:to bucket) (t/seconds max-seconds))]
+    (and (= (:owner bucket) (:owner event))
+         (t/before? from (:timestamp event))
+         (t/before? (:timestamp event) to))))
+
+(defn init-bucket [event]
+  {:count 1
+   :owner (:owner event)
+   :from (:timestamp event)
+   :to (:timestamp event)
+   :events (list event)})
+
+(defn min-ts [t1 t2]
+  (if (t/before? t1 t2) t1 t2))
+
+(defn max-ts [t1 t2]
+  (if (t/before? t1 t2) t2 t1))
+
+(defn bucket-append [bucket event]
+  (-> (update bucket :count inc)
+      (update :from min-ts (:timestamp event))
+      (update :to max-ts (:timestamp event))
+      (update :events conj event)))
+
+(defn timeline-append [comp timeline event]
+  (let [previous (first timeline)]
+    (if (and (map? previous)
+             (same-bucket? previous event))
+        (cons (bucket-append previous event)
+              (rest timeline))
+        (cons (init-bucket event) timeline))))
+
+(defn bucketize-events
+  [events comp]
+  (let [events (sort-by (juxt :owner :timestamp :event_type) events)
```

Comment:
There is an hidden state while performing `bucketize-events`. If those are not sorted only by timestamp the result might not be bucketized as expected.

Let `e1`, ... , `en` events, with timestamp `t1`, ..., `tn`. Let just say all event occur each at the same interval (2/3 of `max-seconds`) then `bucketize-events` will end up with only one bucket. But if the order is `e1`,`e3`...`e(n/2)`,`e(n/2 + 2)` ... the there will be n buckets.

Also another concern is that we might want to only use another single timestamp in the bucket with some tolerance. That way it would be possible to put an event in the correct bucket in a list of buckets.